token-lending/program/src/state/reserve.rs

```diff
@@ -544,12 +723,240 @@ impl Pack for Reserve {
+            let collateral_token_supply = Decimal::from(total_liquidity)
+                .try_mul(Rate::from_scaled_val(collateral_exchange_rate))?
+                .try_floor_u64()?;
+
+            let reserve = Reserve {
+                collateral: ReserveCollateral {
+                    mint_total_supply: collateral_token_supply,
+                    ..ReserveCollateral::default()
+                },
+                liquidity: ReserveLiquidity {
+                    available_amount: total_liquidity,
+                    ..ReserveLiquidity::default()
+                },
+                config: ReserveConfig {
+                    liquidation_threshold,
+                    liquidation_bonus,
+                    ..ReserveConfig::default()
+                },
+                ..Reserve::default()
+            };
+
+            let conversion_rate = Decimal::from_scaled_val(token_conversion_rate);
+            let liquidate_result = reserve.liquidate_obligation(
+                &obligation,
+                liquidate_amount,
+                &Pubkey::default(),
+                MockConverter(conversion_rate)
+            );
+
+            let collateral_exchange_rate = reserve.collateral_exchange_rate()?;
+            let obligation_ltv = loan_to_value_ratio(&obligation, conversion_rate, collateral_exchange_rate)?;
+            if obligation_ltv <= Rate::from_percent(liquidation_threshold) {
+                assert_eq!(
+                    liquidate_result.unwrap_err(),
+                    LendingError::HealthyObligation.into()
+                );
+            } else {
+                let liquidate_result = liquidate_result.unwrap();
+                assert!(
+                    Decimal::from(liquidate_result.withdraw_amount) <=
+                        liquidity_in_other_collateral(
+                            liquidate_result.integer_repay_amount,
+                            collateral_exchange_rate,
+                            conversion_rate,
+                        )?
+                );
+                assert!(
+                    Decimal::from(liquidate_result.integer_repay_amount) <=
+                        obligation.borrowed_liquidity_wads.try_mul(Rate::from_percent(LIQUIDATION_CLOSE_FACTOR))?
+                );
+                assert!(
+                    Decimal::from(liquidate_result.bonus_amount) <=
+                        Decimal::from(liquidate_result.withdraw_amount).try_mul(Rate::from_percent(liquidation_bonus))?
+                );
+                assert!(
+                    liquidate_result.withdraw_amount + liquidate_result.bonus_amount <=
+                        obligation.deposited_collateral_tokens
+                );
+            }
+        }
+
+        #[test]
+        fn allowed_borrow_for_collateral(
+            collateral_amount in 0..=u32::MAX as u64,
+            collateral_exchange_rate in PERCENT_SCALER..=5 * WAD,
+            token_conversion_rate in 0..=u64::MAX as u128,
```

Comment:
Should this be `1..=u64::MAX` like in the other test?

Reply:
Good question. I opted to change the other to `0..=u64::MAX` with the mentality that wider ranges are better for proptests. Thanks for calling out the inconsistency

Reply:
Err, not sure how the other passed with `0..=u64::MAX` because it's not valid. I'll follow up in another PR to make these consistent in starting with `1` as the minimum exchange rate

Reply:
It's kind of worrying that the proptests don't automatically test the limits, we might have to add some proptest utilities to make sure it always goes for the top and bottom of all ranges.  Not sure if you remember, but I ran into the same issue with the swap proptests.

Reply:
Yes, I remember and agree :/